### PR TITLE
Add double-bracket strings

### PR DESCRIPTION
**Pull request by max-sixty**

### CODE REVIEW

**Comment by max-sixty** [comment]
Would close https://github.com/prql/prql/issues/721. I need to add to the docs.

I had previously implemented an `rs"foo"` string. But actually I think the double-bracket approach is fairly standard, and more closely aligns with the current problem. Definitely open to feedback though.

File: prql-compiler/src/prql.pest [hunk 1]
```diff
@@ -136,9 +136,16 @@ operator_compare = ${ "==" | "!=" | ">" | "<" | ">=" | "<=" }
 operator_logical = ${ ("and" | "or") ~ &WHITESPACE }
 operator_coalesce = ${ "??" }
 
-s_string = ${ "s" ~ opening_quote ~ ( interpolate_string_inner | ( "{" ~ expr_call ~ "}" ))* ~ POP }
-f_string = ${ "f" ~ opening_quote ~ ( interpolate_string_inner | ( "{" ~ expr_call ~ "}" ))* ~ POP }
-interpolate_string_inner = { ( !( PEEK | "{" ) ~ ANY )+ }
+// If we have lots more string prefixes then wwe could just have a type
```
Comment (max-sixty) [inline prql-compiler/src/prql.pest]:
```suggestion
// If we have lots more string prefixes then we could just have a type
```

**Comment by max-sixty** [comment]
Currently the SQL formatter is changing `{?foo}` to `( ? foo }`. I need to think about the best way of dealing with this — is there some generalized way (i.e. all valid across dialects) of surrounding the expression so that the SQL formatter doesn't mangle it, but it's still valid?

Otherwise we'll have to replace it with something and then swap it back after the formatting, which would be a bit inelegant.

**Comment by max-sixty** [comment]
Possibly the way to do this is to surround double-bracket strings with dialect-specific "quotes" — e.g. for MySQL & BigQuery, we use backticks; for others we use double-quotes.

I can't see an escape hatch like we do for jinja expressions (https://github.com/prql/prql/blob/0.2.0/prql-compiler/src/sql/translator.rs#L46).

**Comment by mklopets** [comment]
> Currently the SQL formatter is changing `{?foo}` to `( ? foo }`

fyi, just as a related data point, our PRQL pipeline already does this kind of replacing post-compilation as we use postgres's `@?` inside of an S-string, which gets mangled to `@ ?` (i.e. invalid sql)

**Comment by max-sixty** [comment]
Thanks for that case @mklopets 

I increasingly think we should parse to some representation of "raw ident", e.g. backticks, and then based on the dialect convert to either backticks or double-quotes.

**Comment by max-sixty** [comment]
> I increasingly think we should parse to some representation of "raw ident", e.g. backticks, and then based on the dialect convert to either backticks or double-quotes.

I'm planning to pursue this, combined with the approach in #809.

So:

```elm
from foo
select `{{bar}}`
```

becomes, for dialects that use double-quotes to surround "raw identifiers":

```sql
select "{bar}" from foo
```

or for those that use backticks:

```sql
select `{bar}` from foo
```

The only change here would be to compile `{{bar}}` into `{bar}` — i.e. double brackets become single brackets with no evaluation of the contained expression. Currently is an error since it attempts to evaluate `{bar}` as a PRQL expression.

---

Edit: for any future travelers, I didn't end up pursuing this exact approach — the s-string approach seems to cover the cases, with fewer additional language features.

**Comment by max-sixty** [comment]
Not sure what's going on with the commit history but it'll be squashed...

Added some docs too; lmk if anything is unclear!

**Pull request merged**